src/ripple/app/ledger/impl/LedgerReplayer.cpp

```diff
@@ -0,0 +1,284 @@
+                                << totalNumLedgers
+                                << " ledgers merged into an existing task.";
+                return;
+            }
+        }
+        JLOG(j_.info()) << "Replay " << totalNumLedgers
+                        << " ledgers. Finish ledger hash "
+                        << parameter.finishHash;
+
+        auto i = skipLists_.find(parameter.finishHash);
+        if (i != skipLists_.end())
+        {
+            skipList = i->second.lock();
+            if (!skipList)
+            {
+                skipLists_.erase(i);
+            }
+        }
+
+        if (!skipList)
+        {
+            skipList = std::make_shared<SkipListAcquire>(
+                app_,
+                inboundLedgers_,
+                parameter.finishHash,
+                peerSetBuilder_->build());
+            skipLists_.emplace(parameter.finishHash, skipList);
+            newSkipList = true;
+        }
+
+        task = std::make_shared<LedgerReplayTask>(
+            app_, inboundLedgers_, *this, skipList, std::move(parameter));
+        tasks_.push_back(task);
+    }
+
+    if (newSkipList)
+        skipList->init(1);
+    // task init after skipList init, could save a timeout
+    task->init();
+}
+
+void
+LedgerReplayer::createDeltas(std::shared_ptr<LedgerReplayTask> task)
+{
+    {
+        // TODO for use cases like Consensus (i.e. totalLedgers = 1 or small):
+        // check if the last closed or validated ledger l the local node has
+        // is in the skip list and is an ancestor of parameter.startLedger
+        // that has to be downloaded, if so expand the task to start with l.
+    }
+
+    auto const& parameter = task->getTaskParameter();
+    JLOG(j_.trace()) << "Creating " << parameter.totalLedgers - 1 << " deltas";
+    if (parameter.totalLedgers > 1)
+    {
+        auto skipListItem = std::find(
+            parameter.skipList.begin(),
+            parameter.skipList.end(),
+            parameter.startHash);
+        if (skipListItem == parameter.skipList.end() ||
+            ++skipListItem == parameter.skipList.end())
+        {
+            JLOG(j_.error()) << "Task parameter error when creating deltas "
+                             << parameter.finishHash;
+            return;
+        }
+
+        for (std::uint32_t seq = parameter.startSeq + 1;
+             seq <= parameter.finishSeq &&
+             skipListItem != parameter.skipList.end();
+             ++seq, ++skipListItem)
+        {
+            bool newDelta = false;
+            std::shared_ptr<LedgerDeltaAcquire> delta = {};
+            {
+                std::lock_guard<std::mutex> lock(mtx_);
+                auto i = deltas_.find(*skipListItem);
+                if (i != deltas_.end())
+                {
+                    delta = i->second.lock();
+                    if (!delta)
+                    {
+                        deltas_.erase(i);
+                    }
```

Comment:
I don't think you have to erase here. I think you can just do:
```
if (!delta)
{
    delta = std::make_shared<LedgerDeltaAcquire>(
        app_,
        inboundLedgers_,
        *skipListItem,
        seq,
        peerSetBuilder_->build());
        newDelta = true;
}
```
And then get rid of the `if` starting on line 162

Reply:
same as the `SkipList` case.

Reply:
Changed after tests.